File: r/causalCalibration/R/bundles.R
#' Construct a calibration bundle from vectors
#'
#' @param predictions Numeric vector of treatment-effect predictions.
#' @param treatment Numeric binary vector of treatment assignments.
#' @param outcome Numeric vector of observed outcomes.
#' @param mu0 Optional control outcome regression estimates.
#' @param mu1 Optional treated outcome regression estimates.
#' @param outcome_mean Optional marginal outcome estimates.
#' @param propensity Optional propensity estimates.
#' @param sample_weight Optional observation weights.
#'
#' @return An object of class `"calibration_bundle"`.
#' @export
calibration_bundle <- function(
  predictions,
  treatment,
  outcome,
  mu0 = NULL,
  mu1 = NULL,
  outcome_mean = NULL,
  propensity = NULL,
  sample_weight = NULL
) {
  structure(
    list(
      predictions = .cc_as_numeric_vector(predictions, "predictions"),
      treatment = .cc_as_numeric_vector(treatment, "treatment"),
      outcome = .cc_as_numeric_vector(outcome, "outcome"),
      mu0 = if (is.null(mu0)) NULL else .cc_as_numeric_vector(mu0, "mu0"),
      mu1 = if (is.null(mu1)) NULL else .cc_as_numeric_vector(mu1, "mu1"),
      outcome_mean = if (is.null(outcome_mean)) NULL else .cc_as_numeric_vector(outcome_mean, "outcome_mean"),
      propensity = if (is.null(propensity)) NULL else .cc_as_numeric_vector(propensity, "propensity"),
      sample_weight = if (is.null(sample_weight)) NULL else .cc_as_numeric_vector(sample_weight, "sample_weight")
    ),
    class = "calibration_bundle"
  )
}

#' Construct a calibration bundle from a data frame
#'
#' @param data A `data.frame`.
#' @param predictions,treatment,outcome,mu0,mu1,outcome_mean,propensity,sample_weight Column names.
#'
#' @return An object of class `"calibration_bundle"`.
#' @export
calibration_bundle_from_data_frame <- function(
  data,
  predictions = "predictions",
  treatment = "treatment",
  outcome = "outcome",
  mu0 = "mu0",
  mu1 = "mu1",
  outcome_mean = "outcome_mean",
  propensity = "propensity",
  sample_weight = "sample_weight"
) {
  if (!is.data.frame(data)) {
    stop("`data` must be a data.frame.", call. = FALSE)
  }
  calibration_bundle(
    predictions = data[[predictions]],
    treatment = data[[treatment]],
    outcome = data[[outcome]],
    mu0 = if (is.null(mu0)) NULL else data[[mu0]],
    mu1 = if (is.null(mu1)) NULL else data[[mu1]],
    outcome_mean = if (is.null(outcome_mean)) NULL else data[[outcome_mean]],
    propensity = if (is.null(propensity)) NULL else data[[propensity]],
    sample_weight = if (is.null(sample_weight)) NULL else data[[sample_weight]]
  )
}

#' Construct a cross-fit bundle from vectors
#'
#' @inheritParams calibration_bundle
#' @param fold_predictions Numeric matrix of fold-specific predictions.
#' @param fold_ids Optional integer vector identifying the out-of-fold column for each observation.
#'
#' @return An object of class `"crossfit_bundle"`.
#' @export
crossfit_bundle <- function(
  predictions,
  fold_predictions,
  treatment,
  outcome,
  mu0 = NULL,
  mu1 = NULL,
  outcome_mean = NULL,
  propensity = NULL,
  sample_weight = NULL,
  fold_ids = NULL
) {
  base <- calibration_bundle(
    predictions = predictions,
    treatment = treatment,
    outcome = outcome,
    mu0 = mu0,
    mu1 = mu1,
    outcome_mean = outcome_mean,
    propensity = propensity,
    sample_weight = sample_weight
  )
  structure(
    c(
      base,
      list(
        fold_predictions = .cc_as_matrix(fold_predictions, "fold_predictions"),
        fold_ids = if (is.null(fold_ids)) NULL else as.integer(fold_ids)
      )
    ),
    class = c("crossfit_bundle", "calibration_bundle")
  )
}

#' Construct a cross-fit bundle from a data frame
#'
#' @param data A `data.frame`.
#' @param fold_prediction_columns Character vector naming fold-specific prediction columns.
#' @param fold_ids Optional fold-id column name.
#' @inheritParams calibration_bundle_from_data_frame
#'
#' @return An object of class `"crossfit_bundle"`.
#' @export
crossfit_bundle_from_data_frame <- function(
  data,
  fold_prediction_columns,
  fold_ids = NULL,
  predictions = "predictions",
  treatment = "treatment",
  outcome = "outcome",
  mu0 = "mu0",
  mu1 = "mu1",
  outcome_mean = "outcome_mean",
  propensity = "propensity",
  sample_weight = "sample_weight"
) {
  if (!is.data.frame(data)) {
    stop("`data` must be a data.frame.", call. = FALSE)
  }
  crossfit_bundle(
    predictions = data[[predictions]],
    fold_predictions = as.matrix(data[, fold_prediction_columns, drop = FALSE]),
    treatment = data[[treatment]],
    outcome = data[[outcome]],
    mu0 = if (is.null(mu0)) NULL else data[[mu0]],
    mu1 = if (is.null(mu1)) NULL else data[[mu1]],
    outcome_mean = if (is.null(outcome_mean)) NULL else data[[outcome_mean]],
    propensity = if (is.null(propensity)) NULL else data[[propensity]],
    sample_weight = if (is.null(sample_weight)) NULL else data[[sample_weight]],
    fold_ids = if (is.null(fold_ids)) NULL else data[[fold_ids]]
  )
}

#' Validate a bundle object
#'
#' @param bundle A calibration or cross-fit bundle.
#' @param tolerance Alignment tolerance for cross-fit validation.
#'
#' @return A named list summarizing the validated bundle.
#' @export
validate_bundle <- function(bundle, tolerance = 1e-8) {
  if (inherits(bundle, "crossfit_bundle")) {
    return(
      validate_crossfit_bundle(
        predictions = bundle$predictions,
        fold_predictions = bundle$fold_predictions,
        fold_ids = bundle$fold_ids,
        tolerance = tolerance
      )
    )
  }
  if (!inherits(bundle, "calibration_bundle")) {
    stop("`bundle` must inherit from `calibration_bundle` or `crossfit_bundle`.", call. = FALSE)
  }
  .cc_validate_same_length(
    length(bundle$predictions),
    treatment = bundle$treatment,
    outcome = bundle$outcome
  )
  if (!is.null(bundle$mu0)) {
    .cc_validate_same_length(length(bundle$predictions), mu0 = bundle$mu0)
  }
  if (!is.null(bundle$mu1)) {
    .cc_validate_same_length(length(bundle$predictions), mu1 = bundle$mu1)
  }
  if (!is.null(bundle$outcome_mean)) {
    .cc_validate_same_length(length(bundle$predictions), outcome_mean = bundle$outcome_mean)
  }
  if (!is.null(bundle$propensity)) {
    .cc_validate_same_length(length(bundle$predictions), propensity = bundle$propensity)
  }
  if (!is.null(bundle$sample_weight)) {
    .cc_validate_same_length(length(bundle$predictions), sample_weight = bundle$sample_weight)
  }
  list(n_obs = length(bundle$predictions))
}

#' Fit a calibrator from a bundle
#'
#' @param bundle A `calibration_bundle`.
#' @param ... Additional arguments passed to [fit_calibrator()].
#'
#' @return A fitted calibrator.
#' @export
fit_bundle_calibrator <- function(bundle, ...) {
  if (!inherits(bundle, "calibration_bundle")) {
    stop("`bundle` must inherit from `calibration_bundle`.", call. = FALSE)
  }
  fit_calibrator(
    predictions = bundle$predictions,
    treatment = bundle$treatment,
    outcome = bundle$outcome,
    mu0 = bundle$mu0,
    mu1 = bundle$mu1,
    outcome_mean = bundle$outcome_mean,
    propensity = bundle$propensity,
    sample_weight = bundle$sample_weight,
    ...
  )
}

#' Fit a cross-calibrator from a cross-fit bundle
#'
#' @param bundle A `crossfit_bundle`.
#' @param ... Additional arguments passed to [fit_cross_calibrator()].
#'
#' @return A fitted cross-calibrator.
#' @export
fit_bundle_cross_calibrator <- function(bundle, ...) {
  if (!inherits(bundle, "crossfit_bundle")) {
    stop("`bundle` must inherit from `crossfit_bundle`.", call. = FALSE)
  }
  fit_cross_calibrator(
    predictions = bundle$predictions,
    fold_predictions = bundle$fold_predictions,
    fold_ids = bundle$fold_ids,
    treatment = bundle$treatment,
    outcome = bundle$outcome,
    mu0 = bundle$mu0,
    mu1 = bundle$mu1,
    outcome_mean = bundle$outcome_mean,
    propensity = bundle$propensity,
    sample_weight = bundle$sample_weight,
    ...
  )
}

#' @export
print.calibration_bundle <- function(x, ...) {
  cat("<calibration_bundle>\n")
  cat(sprintf("  n_obs: %s\n", length(x$predictions)))
  invisible(x)
}

#' @export
print.crossfit_bundle <- function(x, ...) {
  cat("<crossfit_bundle>\n")
  cat(sprintf("  n_obs: %s\n", length(x$predictions)))
  cat(sprintf("  n_folds: %s\n", ncol(x$fold_predictions)))
  invisible(x)
}

File: python/src/causal_calibration/bundles.py
"""Workflow bundle helpers for external prediction pipelines."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Mapping

from ._utils import (
    as_matrix_rows,
    as_optional_vector,
    as_vector,
    get_column,
    validate_same_length,
)


def _frame_column(frame: Any, name: str) -> Any:
    if isinstance(frame, Mapping):
        return frame[name]
    return frame[name]


@dataclass
class CalibrationBundle:
    predictions: list[float]
    treatment: list[float]
    outcome: list[float]
    mu0: list[float] | None = None
    mu1: list[float] | None = None
    outcome_mean: list[float] | None = None
    propensity: list[float] | None = None
    sample_weight: list[float] | None = None

    @classmethod
    def from_mapping(cls, mapping: Mapping[str, object]) -> "CalibrationBundle":
        return cls(
            predictions=as_vector(get_column(mapping, "predictions"), "predictions"),
            treatment=as_vector(get_column(mapping, "treatment"), "treatment"),
            outcome=as_vector(get_column(mapping, "outcome"), "outcome"),
            mu0=None if get_column(mapping, "mu0", required=False) is None else as_vector(get_column(mapping, "mu0"), "mu0"),
            mu1=None if get_column(mapping, "mu1", required=False) is None else as_vector(get_column(mapping, "mu1"), "mu1"),
            outcome_mean=None
            if get_column(mapping, "outcome_mean", required=False) is None
            else as_vector(get_column(mapping, "outcome_mean"), "outcome_mean"),
            propensity=None
            if get_column(mapping, "propensity", required=False) is None
            else as_vector(get_column(mapping, "propensity"), "propensity"),
            sample_weight=None
            if get_column(mapping, "sample_weight", required=False) is None
            else as_vector(get_column(mapping, "sample_weight"), "sample_weight"),
        )

    @classmethod
    def from_frame(
        cls,
        frame: Any,
        *,
        predictions: str = "predictions",
        treatment: str = "treatment",
        outcome: str = "outcome",
        mu0: str | None = "mu0",
        mu1: str | None = "mu1",
        outcome_mean: str | None = "outcome_mean",
        propensity: str | None = "propensity",
        sample_weight: str | None = "sample_weight",
    ) -> "CalibrationBundle":
        return cls(
            predictions=as_vector(_frame_column(frame, predictions), "predictions"),
            treatment=as_vector(_frame_column(frame, treatment), "treatment"),
            outcome=as_vector(_frame_column(frame, outcome), "outcome"),
            mu0=None if mu0 is None else as_vector(_frame_column(frame, mu0), "mu0"),
            mu1=None if mu1 is None else as_vector(_frame_column(frame, mu1), "mu1"),
            outcome_mean=None if outcome_mean is None else as_vector(_frame_column(frame, outcome_mean), "outcome_mean"),
            propensity=None if propensity is None else as_vector(_frame_column(frame, propensity), "propensity"),
            sample_weight=None if sample_weight is None else as_vector(_frame_column(frame, sample_weight), "sample_weight"),
        )

    def validate(self) -> dict[str, int]:
        validate_same_length(
            len(self.predictions),
            treatment=self.treatment,
            outcome=self.outcome,
        )
        if self.mu0 is not None:
            validate_same_length(len(self.predictions), mu0=self.mu0)
        if self.mu1 is not None:
            validate_same_length(len(self.predictions), mu1=self.mu1)
        if self.outcome_mean is not None:
            validate_same_length(len(self.predictions), outcome_mean=self.outcome_mean)
        if self.propensity is not None:
            validate_same_length(len(self.predictions), propensity=self.propensity)
        if self.sample_weight is not None:
            validate_same_length(len(self.predictions), sample_weight=self.sample_weight)
        return {"n_obs": len(self.predictions)}

    def fit_calibrator(self, **kwargs: Any) -> Any:
        from .core import fit_calibrator

        payload: dict[str, Any] = {
            "predictions": self.predictions,
            "treatment": self.treatment,
            "outcome": self.outcome,
            "mu0": self.mu0,
            "mu1": self.mu1,
            "outcome_mean": self.outcome_mean,
            "propensity": self.propensity,
            "sample_weight": self.sample_weight,
        }
        payload.update(kwargs)
        return fit_calibrator(**payload)


@dataclass
class CrossFitBundle(CalibrationBundle):
    fold_predictions: list[list[float]] | None = None
    fold_ids: list[int] | None = None

    @classmethod
    def from_mapping(cls, mapping: Mapping[str, object]) -> "CrossFitBundle":
        base = CalibrationBundle.from_mapping(mapping)
        return cls(
            predictions=base.predictions,
            treatment=base.treatment,
            outcome=base.outcome,
            mu0=base.mu0,
            mu1=base.mu1,
            outcome_mean=base.outcome_mean,
            propensity=base.propensity,
            sample_weight=base.sample_weight,
            fold_predictions=as_matrix_rows(get_column(mapping, "fold_predictions"), "fold_predictions"),
            fold_ids=None
            if get_column(mapping, "fold_ids", required=False) is None
            else [int(value) for value in as_vector(get_column(mapping, "fold_ids"), "fold_ids")],
        )

    @classmethod
    def from_frame(
        cls,
        frame: Any,
        *,
        fold_prediction_columns: list[str],
        fold_ids: str | None = None,
        **kwargs: Any,
    ) -> "CrossFitBundle":
        base = CalibrationBundle.from_frame(frame, **kwargs)
        fold_matrix = as_matrix_rows(
            [[_frame_column(frame, column)[index] for column in fold_prediction_columns] for index in range(len(base.predictions))],
            "fold_predictions",
        )
        fold_id_vector = None
        if fold_ids is not None:
            fold_id_vector = [int(value) for value in as_vector(_frame_column(frame, fold_ids), "fold_ids")]
        return cls(
            predictions=base.predictions,
            treatment=base.treatment,
            outcome=base.outcome,
            mu0=base.mu0,
            mu1=base.mu1,
            outcome_mean=base.outcome_mean,
            propensity=base.propensity,
            sample_weight=base.sample_weight,
            fold_predictions=fold_matrix,
            fold_ids=fold_id_vector,
        )

    def validate(self, tolerance: float = 1e-8) -> dict[str, float]:
        from .core import validate_crossfit_bundle

        if self.fold_predictions is None:
            raise ValueError("`fold_predictions` must be supplied for a CrossFitBundle.")
        return validate_crossfit_bundle(
            predictions=self.predictions,
            fold_predictions=self.fold_predictions,
            fold_ids=self.fold_ids,
            tolerance=tolerance,
        )

    def fit_cross_calibrator(self, **kwargs: Any) -> Any:
        from .core import fit_cross_calibrator

        if self.fold_predictions is None:
            raise ValueError("`fold_predictions` must be supplied for a CrossFitBundle.")
        payload: dict[str, Any] = {
            "predictions": self.predictions,
            "fold_predictions": self.fold_predictions,
            "fold_ids": self.fold_ids,
            "treatment": self.treatment,
            "outcome": self.outcome,
            "mu0": self.mu0,
            "mu1": self.mu1,
            "outcome_mean": self.outcome_mean,
            "propensity": self.propensity,
            "sample_weight": self.sample_weight,
        }
        payload.update(kwargs)
        return fit_cross_calibrator(**payload)
